sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala

```diff
@@ -1116,6 +1116,18 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     checkAnswer(sql("SELECT a.`c.b`, `b.$q`[0].`a@!.q`, `q.w`.`w.i&`[0] FROM t"), Row(1, 1, 1))
   }
 
+  test("SPARK-10656: select(df(*)) fails when a column has special characters") {
+    // user should backtick the name if the column with special character '.'
+    // otherwise it's hard to differentiate the intention of "a.b" and "g.f"
+    val df = sqlContext.read.json(sqlContext.sparkContext.makeRDD(
+      """{"a.b": 10, "d": 11, "f": {"g": 12} }""" :: Nil))
+    checkAnswer(df.select("f.g"), Row(12))
+    checkAnswer(df.select("`a.b`"), Row(10))
+    checkAnswer(df.select("*"), Row(10, 11, Row(12)))
+    checkAnswer(df.withColumnRenamed("f", "h").select("h"), Row(Row(12)))
+    checkAnswer(df.withColumnRenamed("f", "h").select("`a.b`"), Row(10))
```

Comment:
Just add more test to cover the if branch you mentioned. and the below items would fail if without this patch.
``` scala

    checkAnswer(df.select(df("*")), Row(10, 11, Row(12)))
    checkAnswer(df.withColumnRenamed("f", "h").select("h"), Row(Row(12)))
    checkAnswer(df.withColumnRenamed("f", "f").select("f"), Row(Row(12)))
    checkAnswer(df.withColumnRenamed("`a.b`", "s").select("s"), Row(10))
    checkAnswer(df.withColumnRenamed("f", "h").select("`a.b`"), Row(10))
```